File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    answer = True
    
    phone_book.sort()
    # print(phone_book)
    for i in range(len(phone_book)-1):
        if len(phone_book[i]) > len(phone_book[i+1]):
            if phone_book[i+1] == phone_book[i][:len(phone_book[i+1])]:
                return False
        elif len(phone_book[i]) < len(phone_book[i+1]):
            if phone_book[i] == phone_book[i+1][:len(phone_book[i])]:
                return False
        else:
             if phone_book[i] == phone_book[i+1]:
                return False
            
    return answer

# 접두어 접두사?

# 배열안에 문자열 있을때 길이순으로 정렬하려면 어떻게하지
#배열 하나 더만들어서 ㅣㅇㄴ덱스 순으로 넣으면 될거같긴함

# 넣은다음에 해당 길이만큼 존재하는지 슬라이스해서? 이중 ofor문될거같은데 n^2 어떻게 더 줄일까..
# 

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.cpp
#include <string>
#include <vector>
#include <iostream>
#include <unordered_map>
using namespace std;

bool solution(vector<string> phone_book) {
    bool answer = true;
    unordered_map<string, int> m;
    
    for (string s : phone_book){
        m[s] = 0;
    }
    
    for(string s : phone_book){
        string prefix;
        for(int i=0; i<s.size()-1; i++){
            prefix += s[i];
            if(m.find(prefix) != m.end()){
                return false;
            }
        }
    }
    
    return answer;
}


// 1.맵에 다넣기 -> phone_book for 문 안에 -> s길이-1만큼 for문하고 map에 존재하면 반환
// 2.정렬 -> 사전순 -> for -1만큼해서 앞에값하고 비교했는데 앞에 길이만큼 substr로 일치하면 반환false